File: src/FASTCAMPUS/ALGO/D46/P7569.java
package FASTCAMPUS.ALGO.D46;
import java.io.*;
import java.util.*;

public class P7569 {
    static int M,N,H;
    static int [][][] table;
    static int [][][] visited;
    static int [][] dirTable = {{0, -1, 0}, {0, 1, 0}, {0, 0, -1}, {0, 0, 1}, {-1, 0, 0}, {1, 0, 0}};
    static void input() {
        FastReader scan = new FastReader();
        M = scan.nextInt();
        N = scan.nextInt();
        H = scan.nextInt();
        table = new int [H][N][M];
        visited = new int [H][N][M];
        for(int h = 0; h < H; h ++){
            for(int n = 0; n < N ; n ++){
                for(int m = 0; m < M; m ++){
                    table[h][n][m] = scan.nextInt();
                    if(table[h][n][m] == -1){
                        visited[h][n][m] = -2;    
                    } else {
                        visited[h][n][m] = -1;
                    }
                    
                }
            }
        }
    }
    public static void bfs(int h, int n, int m, int d){
        //let Q = [[h,n,m,d]];
        LinkedList <int[]> Q = new LinkedList<int[]>();
        Q.add(new int[]{h,n,m,d});
        while(!Q.isEmpty()){
          int [] pop  = Q.poll();
          for(int i = 0; i < dirTable.length; i ++){
              int[] dir = dirTable[i];
              int nh = pop[0] + dir[0];
              int nn = pop[1] + dir[1];
              int nm = pop[2] + dir[2];
              int nd = pop[3] + 1; 
              if(nh < 0 || nh >= H) continue;
              if(nn < 0 || nn >= N) continue;
              if(nm < 0 || nm >= M) continue;
              //빈칸
              if(visited[nh][nn][nm] == -2) continue;
              if(visited[nh][nn][nm] == -1) {
                visited[nh][nn][nm] = nd;
                Q.add(new int[]{nh, nn, nm, nd});
              } else {
                if(visited[nh][nn][nm] > nd){
                    visited[nh][nn][nm] = nd;
                    Q.add(new int[]{nh, nn, nm, nd});
                }
              }
              
          }
        }        
      }
    public static void pro(){
        for(int h = 0; h < H; h ++){
            for(int n = 0; n < N; n ++){
              for(int m = 0; m < M; m ++){
                if(table[h][n][m] == 1){
                  visited[h][n][m] = 0;
                  bfs(h,n,m,0);
                }
              }
            }
          }
        int max = 0;
        for(int h = 0; h < H; h ++){
            for(int n = 0; n < N; n ++ ){
                for(int m = 0; m < M; m ++){
                    if(visited[h][n][m] == -1){
                    System.out.println(-1);
                    return;
                    }
                    max = Math.max(max , visited[h][n][m]);
                }
            }
        }
        System.out.println(max);
    }
    public static void main(String[] args) {
        input();
        pro();
    }
    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}

File: src/FASTCAMPUS/ALGO/D46/P7569.js
const fs = require('fs');
class Scanner {
  constructor(fileName){
    this.read = fs.readFileSync(fileName).toString().trim().split('\n');
    this.stringToken = [];
  }
  next(){
    if(this.stringToken.length === 0){
      if(this.read.length > 0){
        this.stringToken = this.read.shift().split(" ");
      } else {
        return null;
      }
    } 
    return this.stringToken.shift();
  }

  nextNumber(){
    const next = this.next();
    if(next){
      return Number(next);
    } else {
      return null;
    }
    
  }

  nextLine(){
    if(this.read.length > 0){
      return this.read.shift();
    } else {
      return null;
    }
    
  }
}
const scanner = new Scanner('stdin.txt');
let M, N, H;
let table,visited;
//Z, Y, X
//상 하 좌 우 위 아래
let dirTable = [[0, -1, 0], [0, 1, 0], [0, 0, -1], [0, 0, 1], [-1, 0, 0], [1, 0, 0]];
const input = function (){
  [M, N, H] = scanner.nextLine().split(" ").map(Number);
  table = Array.from({length : H}, () => Array.from({length : N}, () => Array.from({length : M}, () => 0)));
  visited = Array.from({length : H}, () => Array.from({length : N}, () => Array.from({length : M}, () => -1)));
  for(let i = 0; i < H; i ++){
    let infos = scanner.read.splice(0, N);
    table[i] = infos.map(info => info.split(" ").map(Number));
  }
}
const bfs = function (h, n, m, d){
  let Q = [[h,n,m,d]];
  while(Q.length !== 0){
    let [h, n, m, d] = Q.shift();
    dirTable.forEach(([nh, nn, nm]) => {
      if(nh + h < 0 || nh + h >= H) return;
      if(nn + n < 0 || nn + n >= N) return;
      if(nm + m < 0 || nm + m >= M) return;
      if(visited[nh + h][nn + n][nm + m] !== -1) return;
      visited[nh + h][nn + n][nm + m] = d + 1;
      Q.push([nh + h, nn + n,nm + m, d + 1]);
    })
  }
  
}

const pro = function (){
  for(let h = 0; h < H; h ++){
    for(let n = 0; n < N; n ++){
      for(let m = 0; m < M; m ++){
        if(table[h][n][m] === 1){
          visited[h][n][m] = 0;
          bfs(h,n,m,0);
        } else if(table[h][n][m] === -1){
          visited[h][n][m] = 0;
        }
      }
    }
  }
  // console.log(visited);
  let max = 0;
  for(let h = 0; h < H; h ++){
    for(let n = 0; n < N; n ++ ){
      for(let m = 0; m < M; m ++){
        if(visited[h][n][m] === -1){
          console.log(-1);
          return;
        }
        max = Math.max(max , visited[h][n][m]);
      }
    }
  }
  console.log(max);
}


const solve = function(){
  input();
  pro();
}

solve();